### DISCUSSION TITLE
Clasp 3.x roadmap

### DISCUSSION THREAD

**Discussion by sqrrrl** [google/clasp#980] [Announcements]
Hi folks,

Planning the next major release where we can break things a little and am looking for feedback. A few things we want to tackle in the next version:

*  **Drop TypeScript support** -- The built-in TS support leaves a lot to be desired and only supports basic transforms. Moving this outside clasp to a separate build step provides more features and flexibility. We'd like to drop the functionality and encourage people to use tools like Rollup instead.
* **Bring your own credentials** -- Clasp has a built in client ID for auth and allows a custom client ID for a subset of features like using the execution API. We're looking to expand and encourage using your own GCP project for clasp for all use cases.
* **Fix auth in headless environments** -- This current implementation no longer works as it was deprecated and disabled by Google. We need to rework this implementation.
* **Improve error handling** -- There are a few cases where error handling is not what it should be -- failed clones leave behind artifacts, sometimes error messages are cryptic, etc.
* **Better support for multiple environments & profiles** -- Make it easier for those of us that use clasp with multiple accounts or maintain multiple instances of scripts for development vs. production.

Please feel free to add ideas in the discussions or file issues with feature requests :)

**Comment by PopGoesTheWza** [comment]
I have not contributed to this project in a while... for I currently no longer have use for it.

Indeed Typescript support is best left out of Clasp, preferably ealier in the production pipeline, with Clasp only dealing with the JavaScript output.

Just my two cents.

**Reply by senti-mark-monteiro** [reply]
I agree this makes sense 👍 

For reference, here is an example starter project that uses Rollup as a separate build step to handle Typescript: https://github.com/sqrrrl/apps-script-typescript-rollup-starter

**Comment by gfic-limited** [comment]
Make it work seamlessly with gcloud, especially auth.

I don't even remember the magic incantations I did to make it work, and
then at some stage when it broke and I didn't have time to figure it out, I
just cheated by copying and pasting .clasprc.json from one development
machine to another. That's not really good.

Since I am authenticated through gcloud, and have now my apps script tied
to a GCP project, I shouldn't need to worry about any auth with Clasp any
more.

In general, in the big picture, should continue to make apps script and
gcloud interwork, scale and support each other better.

**Comment by red-meadow** [comment]
Correct me if I'm wrong but AFAIK Clasp doesn't support import/export syntax without bundlers. This feature would be very useful.

**Reply by PopGoesTheWza** [reply]
This would imply that Clasp comes with a bundler pipeline... which both users and contributors would have to maintain...

And there will always be need for a different pipeline...

**Comment by ttsukagoshi** [comment]
As an active user of clasp, I maintain some Google Workspace add-ons both internally and externally. This involves using multiple Google accounts; some are Workspace accounts, and others are free, private Gmail accounts. So I definitely support the idea of

> - **Better support for multiple environments & profiles** -- Make it easier for those of us that use clasp with multiple accounts or maintain multiple instances of scripts for development vs. production.

Once you get familiar with the notion of modules in ECMAScript and TypeScript, you feel the urge to use `import`/`export` in Google Apps Script as well; part of me completely agrees with @red-meadow. I think it's the choice of either

1. Fully support TypeScript and modules as a built-in feature of clasp OR
2. Leave TS/module support to other tools like [rollup](https://www.npmjs.com/package/rollup). My understanding from the limited experience with rollup is that it can integrate multiple script files interacting with each other by `import`/`export` statements into a single JS file.

If my understanding is correct, what @sqrrrl is saying in

> - **Drop TypeScript support**

is that the team is trying to steer toward option 2 and concentrate their efforts on improving the auth process around clasp; 👍 +1 from me!

**Comment by PreetM07** [comment]
Hello Guys,

I have been using GAS for like 2 years now. And now my project has 86 files in it.
I started using clasp just now. And frankly, it's very tedious.
Major issue happened when it just rearranged all the files in the GAS editor alphabetically as Explorer after a push.
Also note that "filePushOrder" is not working somehow...
and I am not able to find any potential solution for this issue.
So if someone can please quickly fix or let me know a work around would be a great thing.
Next thing in the roadmap should be something to make helpful to beginners as I see in current roadmap and it looks like it's for novice. Novice will anyhow can do workaround by their skills... I am not sure if I am wrong, but I would like to see clasp growing this way
